### PR TITLE
docs/contributing: Add coding conventions guide

### PR DESCRIPTION
**Pull request by RaduNichita**
Add the coding conventions for Unikraft

### CODE REVIEW

File: content/en/docs/contributing/coding-conventions.md [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
+    break;
+default:
+    /* … */
+    break;
+}
+```
+
+### Goto Statements
+
+Goto labels should not **be indented**.
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:599]:
```suggestion
Goto labels should **not** be indented.
```

File: content/en/docs/contributing/coding-conventions.md [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+```c
+__asm__ __volatile__ (
+    “1:      instr      op, op\n”
+    “        instr2     op, op\n”
+    :::);
+```
+
+* Single-line
+* With / without label
+* `__asm__ __volatile__` vs. `asm volatile`
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:682]: There are TODO-s I guess?

File: content/en/docs/contributing/coding-conventions.md [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+
+1. `UKPLAT_BOOTINFO_MAGIC` 0xB007B0B0 (Boot Bobo)
+1. `FDT_MAGIC 0xd00dfeed` (FDT specification)
+1. `UK_TP_HEADER_MAGIC` 0x64685254 (TRhd)
+1. `UK_TP_DEF_MAGIC` 0x65645054 (TPde)
+1. `CL_UKTLS_SANITY_MAGIC` 0xb0b0f00d (Bobo food)
+1. `0xDEADB0B0`, `0xB0B0CAFE ` (in ukvmem)
+1. `MULTIBOOT_HEADER_MAGIC` 0x1BADB002
+1. `MULTIBOOT_BOOTLOADER_MAGIC` 0x2BADB002
+```
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:822]: What do there backticks close?

File: content/en/docs/contributing/coding-conventions.md [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:823]: Also add that when used in source code, the symbol of the magic number should be used, never the value.

File: content/en/docs/contributing/coding-conventions.md [hunk 5]
```diff
@@ -0,0 +1,823 @@
...
+* When do we break?
+When reaching 80 characters?
+* Where do we break?
+* How do we align after the break?
+
+1. Proposal MR: Break before function name (after function return type)
+1. Proposal MR: Align function arguments after break to open parenthesis
+1. Decision: agreed
+
+TODO: How do we align when we break on conditions, for loops?
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:702]: What was the agreed version here?

File: content/en/docs/contributing/coding-conventions.md [hunk 6]
```diff
@@ -0,0 +1,823 @@
...
+```c
+UK_SYSCALL_R_DEFINE(int, getpeername, int, sock,
+»       »           struct sockaddr *restrict, addr,
+»       »           socklen_t *restrict, addr_len)
+```
+
+For conditions, always aligned to the parenthesis.
+To be voted via chat.
+One breaking rule to rule them all.
+TODO: Wait for Michalis’s insight.
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:715]: Update according to [this](https://discord.com/channels/762976922531528725/857555406187528212/1110131214393806860).

File: content/en/docs/contributing/coding-conventions.md [hunk 7]
```diff
@@ -0,0 +1,823 @@
...
+     c == 12) || \
+    d == 15)
+
+if (a == 0 \
+    && (b == 10 \
+        || c == 12) \
+    || d == 15)
+```
+
+Wait for Michalis’s insight.
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:729]:
```suggestion
```

File: content/en/docs/contributing/coding-conventions.md [hunk 8]
```diff
@@ -0,0 +1,823 @@
...
+1. Proposal MR: Have UK_ASSERT even in the absence of ukdebug
+Decision: Agree (part of the plat re-arch discussion)
+Where possible use the compile time assert (UK_CTASSERT): CT for compile-time
+1. Proposal MP: Guideline when using UK_ASSERT and when using UK_CRASH
+	UK_CRASH is for situations where you can not formalize an assertion
+  	Something like UK_ASSERT(false)
+1. MS: Some should be present in release builds
+UK_BUG_ON should never be removed from the code (with condition)
+The same for UK_CRASH (without condition)
+Change behavior of UK_BUG_ON to always be part of the code, even in release builds
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:758]: Remove the proposals from here, only leave the final decisions.

**Review by StefanJum — CHANGES_REQUESTED** [review]
Thanks a lot @RaduNichita, I left some comments.
It seems there are still a few TODOs left from the document, so also check those out.

File: content/en/docs/contributing/coding-conventions.md [hunk 9]
```diff
@@ -0,0 +1,815 @@
...
+
+* low-level native Unikraft types like `__u64` and `__spinlock`
+* header inclusion guards
+* preprocessor macros to add another level of indirection for macro resolution
+
+{{% alert theme="info" %}}
+Rule of thumb: include `libc` headers first and then Unikraft headers.
+{{% /alert %}}
+
+* Use `uk_nonnull` instead of and `UK_NULL` instead of `__NULL`.
```
Comment (michpappas) [inline content/en/docs/contributing/coding-conventions.md:439]: What are `uk_nonnull`, `UK_NULL` and so on? Can't find any of this in our codebase.

File: content/en/docs/contributing/coding-conventions.md [hunk 10]
```diff
@@ -0,0 +1,815 @@
...
+<UK API>_<COMPONENT>_<NAME>
+```
+
+Unikraft reserves the following API prefixes to describe **functions**, **function-like macros** and **data types** of the Unikraft API. This also applies (CAPITALIZED) for macros:
+
+* `uk_`  -  denotes a public Unikraft API at the library level.
+The implementation is usually found in one of the libraries in `/lib` or defined in the headers in `/include/uk`.
+For example: `uk_alloc()`
+
+* `ukplat_` - denotes a public Unikraft API, whose implementation is usually provided in `/plat` and depends on the platform (e.g., KVM) for which the Unikernel is compiled.
```
Comment (michpappas) [inline content/en/docs/contributing/coding-conventions.md:455]: `ukplat` and `ukarch` are subject to become obsolete by plat-rearch

File: content/en/docs/contributing/coding-conventions.md [hunk 11]
```diff
@@ -0,0 +1,815 @@
...
+* Public is defined in exportsyms.uk
+
+* Private is not as part of exportsyms.uk or part of localsyms.uk.
+
+* If there is no `exportsyms.uk` defined, then all functions are exported.
+{{% /alert %}}
+
+#### Typedefs
+
+*Rule of thumb*: **Do not** use typedefs.
```
Comment (michpappas) [inline content/en/docs/contributing/coding-conventions.md:508]: Additional reasoning here: https://www.kernel.org/doc/html/v4.15/process/coding-style.html#typedefs

**Review by michpappas — CHANGES_REQUESTED** [review]
Hi, this review comes unsolicited, I just wanted to look something up in our guidelines and ended up seeing things here and there. Hope it's okay :)

File: content/en/docs/contributing/coding-conventions.md [hunk 12]
```diff
@@ -0,0 +1,815 @@
+# Unikraft: Coding Conventions (RFC)
+
+Coding conventions are important to maintain consistency and readability in software development, making code easier to understand, debug, review and maintain by multiple developers over time.
+
+These are conventions for writing code and naming and placing things in Unikraft.
+Maintainers would be the ones who judge if there are certain corner cases that a rule doesn’t need to be obeyed.
+Add some examples with extreme cases.
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/en/docs/contributing/coding-conventions.md:7]: Yes, this is a TODO. We can think of 3 scenarios that won't obey the conventions, but still be accepted.

File: content/en/docs/contributing/coding-conventions.md [hunk 13]
```diff
@@ -0,0 +1,815 @@
...
+## Basics
+
+The layout uses the following basic rules.
+You may install [editorconfig](https://editorconfig.org/) as there is a `.editorconfig` file in the top-level dir to automatically enforce these.
+
+* Unikraft uses **tabs + spaces** for indentation, with tabs being 8 characters wide.
+Tabs first, further indentation less than 8 characters are done with spaces (Linux style).
+* Source files are **80 columns** wide.
+    * There are some scenarios where this limit could be passed, due to [readability reasons](https://github.com/torvalds/linux/commit/bdc48fa11e46f867ea4d75fa59ee87a7f48be144).
+    * Add a WARNING (80 characters) and an ERROR (85 characters) level.
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/en/docs/contributing/coding-conventions.md:18]: This should be done by the static code checker part of the CI (i.e. `checkpatch`).

File: content/en/docs/contributing/coding-conventions.md [hunk 14]
```diff
@@ -0,0 +1,815 @@
...
+* Use a separate line to close the comment for actual multi-line comments.
+There is a leading space to align the lines:
+
+```c
+/* This is a multi line comment because it is so incredibly
+ * useful that the contents does not fit into one line
+ */
+```
+
+* Trailing dots are not necessary, unless there is a sentence that you would end with a do.
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/en/docs/contributing/coding-conventions.md]: Is this solved, @RaduNichita? It appears as outdated.

File: content/en/docs/contributing/coding-conventions.md [hunk 15]
```diff
@@ -0,0 +1,815 @@
...
+ * @param[out] p
+ *   Pointer to the newly allocated memory on success, NULL otherwise
+ *
+ * @return
+ *   The number of bytes actually allocated
+ */
+size_t try_alloc(size_t size, void **p);
+```
+
+* Use dots when ending a sentence or when a new sentence starts afterwards.
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/en/docs/contributing/coding-conventions.md:295]: The only reference I could quickly find on using periods is this: https://softwareengineering.stackexchange.com/a/17769/389332

We can make it so that when using a single sentence, the period is suppressed. But for multiple sentence, I see no point in suppressing the period for the last sentence; it's inconsistent and visually weird.

File: content/en/docs/contributing/coding-conventions.md [hunk 16]
```diff
@@ -0,0 +1,815 @@
...
+* The return can be omitted for trivial getter functions:
+
+```c
+/**
+ * Returns the size of the given buffer
+ */
+size_t get_size(void *p);
+```
+
+For return values, use the convention:
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/en/docs/contributing/coding-conventions.md:310]: @michpappas, could you please elaborate on your comment?

File: content/en/docs/contributing/coding-conventions.md [hunk 17]
```diff
@@ -0,0 +1,815 @@
...
+Should be used as HACK:FIXME:
+1. `FIXME`: This works, sort of, but it could be done better. (usually code written in a hurry that needs rewriting)
+1. `BUG`: There is a problem here
+1. `TODO`: No problem, but additional code needs to be written, usually when you are skipping something
+1. `UNDONE`: A reversal or "roll back" of previous code
+1. `OPTIMIZE`: Code can / should be optimized
+
+### Magic Numbers
+
+These are some magic numbers used in the Unikraft code:
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/en/docs/contributing/coding-conventions.md:802]: From my point of view, it's useful to have a list of magic number **names**. Listing (current) values is optional, I see that useful for completeness, but we can skip them if it's confusing.

File: content/en/docs/contributing/coding-conventions.md [hunk 18]
```diff
@@ -0,0 +1,815 @@
...
+
+Generally, use the `unlikely()` macro when checking for error cases (only).
+It is defined in `<uk/essentials.h>`.
+
+When working with the low-level Unikraft API, it is essential to validate parameters using UK_ASSERT when specific parameter ranges need to be enforced. 
+In cases where there is no explicit specification, performing an error check is recommended to ensure parameter validity.
+
+When working with higher-level APIs, parameter validation should be handled by returning appropriate error codes according to the relevant standard (such as POSIX).
+
+### Comment Keywords
```
[the remark this answers is not in the document]
Comment (razvand) [reply content/en/docs/contributing/coding-conventions.md:788]: Same as with the magic numbers, we want to list them to have contributors be aware of them. There is no enforcing, but rather something like: "I'm a contributor, I'm facing situation X. How should I mark this situation in the comments?"

File: content/en/docs/contributing/coding-conventions.md (hunk 9 above)
Comment (razvand) [reply content/en/docs/contributing/coding-conventions.md:439]: We don't want to use underscore-starting symbols, due to potential clashes with libc/compiler symbols. So, instead of `__NULL` we use `UK_NULL`. And instead of `__nonnull` we use `uk_nonnull`.

@RaduNichita, you missed a "`__nonnull`" in the above description.

File: content/en/docs/contributing/coding-conventions.md (hunk 10 above)
Comment (razvand) [reply content/en/docs/contributing/coding-conventions.md:455]: @michpappas, could you suggest a way to rephrase this? Should we drop it altogether?

File: content/en/docs/contributing/coding-conventions.md [hunk 19]
```diff
@@ -0,0 +1,826 @@
...
+
+* Multi line (tab at the beginning, tab between instr. and operands, `\n` at the end)
+* Could also remove tab but that means `\n\t` at the end, which is not nice, if at some point code is changed and label has to be added.
+
+#### Inline Assembly
+
+```c
+asm volatile (
+    “1:      instr      op, op\n”
+    “        instr2     op, op\n”
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:702]:
```suggestion
    "1:      instr      op, op\n"
    "        instr2     op, op\n"
```
Use ASCII

File: content/en/docs/contributing/coding-conventions.md [hunk 20]
```diff
@@ -0,0 +1,826 @@
...
+    :::);
+```
+
+* Single-line
+* With / without label
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:707]:
```suggestion
```

File: content/en/docs/contributing/coding-conventions.md [hunk 21]
```diff
@@ -0,0 +1,826 @@
...
+* Prefer using `asm volatile` over `__asm__ __volatile__` .
+* Aim to add a comment for each inline assembly line, detailing what it does
+    * Ignore for repetitive operations (such as mov-s)
+    * Ignore long lines for inline assembly, when including comments
+    * Aim to have short comments
+    We could also have multi-line comments
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:713]:
```suggestion
* Aim to add a comment for each inline assembly line, detailing what it does.
    * Ignore for repetitive operations (such as mov-s).
    * Ignore long lines for inline assembly, when including comments.
    * Aim to have short comments.
    We could also have multi-line comments.
```

File: content/en/docs/contributing/coding-conventions.md [hunk 22]
```diff
@@ -0,0 +1,826 @@
...
+
+### Breaking Long Lines
+
+1. Break before function name (after function return type)
+1. Align function arguments after break to open parenthesis
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:718]:
```suggestion
1. Break before function name (after function return type).
1. Align function arguments after break to open parenthesis.
```

File: content/en/docs/contributing/coding-conventions.md [hunk 23]
```diff
@@ -0,0 +1,826 @@
...
+
+For functions, align arguments on different line:
+
+```c
+UK_SYSCALL_R_DEFINE(int, getpeername, int, sock,
+»       »           struct sockaddr *restrict, addr,
+»       »           socklen_t *restrict, addr_len)
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:725]:
```suggestion
                    struct sockaddr *restrict, addr,
                    socklen_t *restrict, addr_len)
```

File: content/en/docs/contributing/coding-conventions.md [hunk 24]
```diff
@@ -0,0 +1,826 @@
...
+```
+
+Function name and first argument are atomic.
+
+Prefer the Linux style with the return type on the same line as the function name.
+In case of long functions + first argument, you can break after the return type.
+
+### Preprocessor Macros
+
+This is referenced in [the definitions](#definitions) section for prefixes (i.e. the use of `uk_`, `ukplat_`, `ukarch_` prefixes).
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:744]: The link to `definitions` is broken.

File: content/en/docs/contributing/coding-conventions.md [hunk 25]
```diff
@@ -0,0 +1,826 @@
...
+Recommendation for using macros as (inline) functions?
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:745]: This is still a TODO right?

File: content/en/docs/contributing/coding-conventions.md [hunk 26]
```diff
@@ -0,0 +1,826 @@
...
+* Some macros as functions would still be uppercase (such as `UK_MIN`, `UK_MAX`)
+* Place each argument (that is a variable) between parentheses
+* If using a variable multiple times, it’s better to use a temporary variable declaration in the macro that stores the variable
+
+### UK_ASSERT
+
+1. Have UK_ASSERT even in the absence of ukdebug
+1. UK_CRASH is for situations where you can not formalize an assertion. Something like `UK_ASSERT(false)`.
+1. UK_BUG_ON should never be removed from the code (with condition). The same applies for UK_CRASH (without condition)
+1. Change behavior of UK_BUG_ON to always be part of the code, even in release builds
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:758]:
```suggestion
1. Have `UK_ASSERT` even in the absence of ukdebug
1. `UK_CRASH` is for situations where you can not formalize an assertion.
    Something like `UK_ASSERT(false)`.
1. `UK_BUG_ON` should never be removed from the code (with condition).
    The same applies for `UK_CRASH` (without condition).
1. Change behavior of UK_BUG_ON to always be part of the code, even in release builds
```

**Review by StefanJum — CHANGES_REQUESTED** [review]
See the comments @RaduNichita.

Besides those, there seems to be a few more TODO items.
Also add dots at the end of sentences, even if in a list. I did not add comments to all of them to avoid spam, searched them with `[^.]$`.

File: content/en/docs/contributing/coding-conventions.md (hunk 25 above)
Comment (RaduNichita) [reply content/en/docs/contributing/coding-conventions.md]: I have reordered the bullets bellow

File: content/en/docs/contributing/coding-conventions.md [hunk 27]
```diff
@@ -0,0 +1,828 @@
...
+```
+
+### Assembly
+
+#### Code
+
+Indentation for ASM code:
+
+* Multi line (tab at the beginning, tab between instr. and operands, `\n` at the end)
+* Could also remove tab but that means `\n\t` at the end, which is not nice, if at some point code is changed and label has to be added.
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:696]: I think we should rephrase it to make it more clear, or just remove it.

**Review by StefanJum — CHANGES_REQUESTED** [review]
Looks better @RaduNichita, thank you. Please check the unresolved conversations, if it is the case mark them as resolved, if they need more input from someone please ask for it.

File: content/en/docs/contributing/coding-conventions.md [hunk 28]
```diff
@@ -0,0 +1,843 @@
+---
+title: Unikraft - Coding Conventions
+date: 2023-05-25T14:09:21+09:00
+draft: false
+weight: 500
+---
+
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:7]: Add a `##` header with the title here.

File: content/en/docs/contributing/coding-conventions.md [hunk 29]
```diff
@@ -0,0 +1,843 @@
...
+* \- (0):  Successfully initialized
+* \- (<0): Negative value with error code
+
+Basically, list return values as list items.
+
+### Struct Member Documentation
+
+* Use double asterisk for multi-line comments (/**).
+
+* When having flags, point to defines of macros in the same or another file.
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:330]:
```suggestion
* When having flags, point to definition of macros in the same or another file.
```

File: content/en/docs/contributing/coding-conventions.md [hunk 30]
```diff
@@ -0,0 +1,843 @@
+---
+title: Unikraft - Coding Conventions
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:2]:
```suggestion
title: Coding Conventions
```
We don't need `Unikraft` in the title.

File: content/en/docs/contributing/coding-conventions.md [hunk 31]
```diff
@@ -0,0 +1,843 @@
...
+
+* If using a variable multiple times, it’s better to use a temporary variable declaration in the macro that stores the variable.
+
+### UK_ASSERT
+
+1. Have `UK_ASSERT` even in the absence of ukdebug.
+1. `UK_CRASH` is for situations where you can not formalize an assertion.
+    Something like `UK_ASSERT(false)`.
+1. `UK_BUG_ON` should never be removed from the code (with condition).
+    The same applies for UK_CRASH (without condition).
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:774]:
```suggestion
    The same applies for `UK_CRASH` (without condition).
```

File: content/en/docs/contributing/coding-conventions.md [hunk 32]
```diff
@@ -0,0 +1,843 @@
...
+1. Change behavior of UK_BUG_ON to always be part of the code, even in release builds.
```
Comment (StefanJum) [inline content/en/docs/contributing/coding-conventions.md:775]:
```suggestion
1. Change behavior of `UK_BUG_ON` to always be part of the code, even in release builds.
```

**Review by StefanJum — COMMENTED** [review]
Looks great @RaduNichita, see the few comments, we'll then wait for @razvand and @michpappas to have a look.

**Review by StefanJum — APPROVED** [review]
Looks fine to me @RaduNichita, thanks a lot :1st_place_medal:

**Review by razvand — APPROVED** [review]
Thanks, @RaduNichita.

Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
